### ISSUE TITLE
Release?

### ISSUE THREAD

**Issue by zvezdochiot** [agl/jbig2enc#106]
New release: 0.30 "accurate":
* neat symbolic threshold 0.92
* make the `weight` parameter adjustable
* new BW threshold = 128
* standalone mode
* set DPI (`-D`)
* leptonica >= 1.83
* python3

Good luck. :smile:

**Comment by zdenop** [comment]
I am planning 1.0 🎈. The project is mature enough + we changed the name of python script for pdf creation.
Things ToDo:
- fix [build_system](https://github.com/agl/jbig2enc/issues?q=is%3Aissue+is%3Aopen+label%3Abuild_system) issues 
- implement clang-format for automated code formatting (--style google)

**Comment by zvezdochiot** [comment]
@zdenop say:
>I am planning 1.0

Good. But now you can 0.30 “accurate” for current changes. They are significant. And leave 1.00 “build” for the build system and so on.

**Comment by DanBloomberg** [comment]
@zdenop Before you do even a 0.3 release, please replace jbig2.cc with the attached file.  The results are much better than the current version at head.  I've tested it on the attached image with cyrillic text.

I confess that I have never issued a PR.  So please do it for me, with this description for the commit:

Replace use of a global threshold with an adaptive threshold algorithm
* keep the -T input parameter to jbig2 as a stub, so that scripts using it will not fail
* tested with sharpening before doing the adaptive map; it has little effect, so not used


![new-gray-2024](https://github.com/user-attachments/assets/3dbeac5f-4615-4169-8457-4910a698290f)

[jbig2.cc.gz](https://github.com/user-attachments/files/18205242/jbig2.cc.gz)

**Comment by zvezdochiot** [comment]
Hi all.

@DanBloomberg  suggested:
```diff
--- jbig2.cc.current    2024-12-20 01:27:20.427554031 +0300
+++ jbig2.cc.dan        2024-12-20 05:47:55.203229459 +0300
@@ -47,9 +47,6 @@
 #define JBIG2_WEIGHT_MIN 0.1f
 #define JBIG2_WEIGHT_MAX 0.9f
 #define JBIG2_WEIGHT_DEF 0.5f
-#define BW_THRESHOLD_MIN 0
-#define BW_THRESHOLD_MAX 255
-#define BW_THRESHOLD_DEF 128
 
 static void
 usage(const char *argv0) {
@@ -61,7 +58,6 @@
   fprintf(stderr, "  -s --symbol-mode: use text region, not generic coder\n");
   fprintf(stderr, "  -t <threshold>: set classification threshold for symbol coder (def: %0.2f)\n", JBIG2_THRESHOLD_DEF);
   fprintf(stderr, "  -w <weight>: set classification weight for symbol coder (def: %0.2f)\n", JBIG2_WEIGHT_DEF);
-  fprintf(stderr, "  -T <bw threshold>: set 1 bpp threshold (def: %d)\n", BW_THRESHOLD_DEF);
   fprintf(stderr, "  -r --refine: use refinement (requires -s: lossless)\n");
   fprintf(stderr, "  -O <outfile>: dump thresholded image as PNG\n");
   fprintf(stderr, "  -2: upsample 2x before thresholding\n");
@@ -215,7 +211,6 @@
   bool pdfmode = false;
   float threshold = JBIG2_THRESHOLD_DEF;
   float weight = JBIG2_WEIGHT_DEF;
-  int bw_threshold = BW_THRESHOLD_DEF;
   bool symbol_mode = false;
   bool refine = false;
   bool up2 = false, up4 = false;
@@ -320,9 +315,11 @@
         return 1;
       }
 
-      if ((threshold < JBIG2_THRESHOLD_MIN) || (threshold > JBIG2_THRESHOLD_MAX)) {
+      if ((threshold < JBIG2_THRESHOLD_MIN) ||
+          (threshold > JBIG2_THRESHOLD_MAX)) {
         fprintf(stderr, "Invalid value for threshold\n");
-        fprintf(stderr, "(must be between %0.2f and %0.2f)\n", JBIG2_THRESHOLD_MIN, JBIG2_THRESHOLD_MAX);
+        fprintf(stderr, "(must be between %0.2f and %0.2f)\n",
+                JBIG2_THRESHOLD_MIN, JBIG2_THRESHOLD_MAX);
         return 10;
       }
       i++;
@@ -340,7 +337,8 @@
 
       if ((weight < JBIG2_WEIGHT_MIN) || (weight > JBIG2_WEIGHT_MAX)) {
         fprintf(stderr, "Invalid value for weight\n");
-        fprintf(stderr, "(must be between %0.2f and %0.2f)\n", JBIG2_WEIGHT_MIN, JBIG2_WEIGHT_MAX);
+        fprintf(stderr, "(must be between %0.2f and %0.2f)\n",
+                JBIG2_WEIGHT_MIN, JBIG2_WEIGHT_MAX);
         return 10;
       }
       i++;
@@ -348,17 +346,7 @@
     }
 
     if (strcmp(argv[i], "-T") == 0) {
-      char *endptr;
-      bw_threshold = strtol(argv[i+1], &endptr, 10);
-      if (*endptr) {
-        fprintf(stderr, "Cannot parse int value: %s\n", argv[i+1]);
-        usage(argv[0]);
-        return 1;
-      }
-      if (bw_threshold < BW_THRESHOLD_MIN || bw_threshold > BW_THRESHOLD_MAX) {
-        fprintf(stderr, "Invalid bw threshold: (%d..%d)\n", BW_THRESHOLD_MIN, BW_THRESHOLD_MAX);
-        return 11;
-      }
+      fprintf(stderr, "Binary thresholds 'T' are no longer used\n");
       i++;
       continue;
     }
@@ -456,7 +444,7 @@
     if (verbose)
       pixInfo(source, "source image:");
 
-    PIX *pixl, *gray, *pixt;
+    PIX *pixl, *gray, *adapt, *pixt;
     if ((pixl = pixRemoveColormap(source, REMOVE_CMAP_BASED_ON_SRC)) == NULL) {
       fprintf(stderr, "Failed to remove colormap from %s\n", argv[i]);
       return 1;
@@ -474,14 +462,14 @@
         fprintf(stderr, "Unsupported input image depth: %d\n", pixl->d);
         return 1;
       }
-      Pix *adapt = pixBackgroundNormSimple(gray, NULL, NULL);
+      adapt = pixCleanBackgroundToWhite(gray, NULL, NULL, 1.0, 90, 190);
       pixDestroy(&gray);
       if (up2) {
-        pixt = pixScaleGray2xLIThresh(adapt, bw_threshold);
+        pixt = pixScaleGray2xLIThresh(adapt, 220);
       } else if (up4) {
-        pixt = pixScaleGray4xLIThresh(adapt, bw_threshold);
+        pixt = pixScaleGray4xLIThresh(adapt, 220);
       } else {
-        pixt = pixThresholdToBinary(adapt, bw_threshold);
+        pixt = pixThresholdToBinary(adapt, 220);
       }
       pixDestroy(&adapt);
     } else {
```

Compromise option (while maintaining threshold controllability):
```diff
--- jbig2.cc.current    2024-12-20 01:27:20.427554031 +0300
+++ jbig2.cc    2024-12-20 06:02:08.707097873 +0300
@@ -49,7 +49,7 @@
 #define JBIG2_WEIGHT_DEF 0.5f
 #define BW_THRESHOLD_MIN 0
 #define BW_THRESHOLD_MAX 255
-#define BW_THRESHOLD_DEF 128
+#define BW_THRESHOLD_DEF 223
 
 static void
 usage(const char *argv0) {
@@ -456,7 +456,7 @@
     if (verbose)
       pixInfo(source, "source image:");
 
-    PIX *pixl, *gray, *pixt;
+    PIX *pixl, *gray, *adapt, *pixt;
     if ((pixl = pixRemoveColormap(source, REMOVE_CMAP_BASED_ON_SRC)) == NULL) {
       fprintf(stderr, "Failed to remove colormap from %s\n", argv[i]);
       return 1;
@@ -474,7 +474,7 @@
         fprintf(stderr, "Unsupported input image depth: %d\n", pixl->d);
         return 1;
       }
-      Pix *adapt = pixBackgroundNormSimple(gray, NULL, NULL);
+      adapt = pixCleanBackgroundToWhite(gray, NULL, NULL, 1.0, 90, 190);
       pixDestroy(&gray);
       if (up2) {
         pixt = pixScaleGray2xLIThresh(adapt, bw_threshold);
```

Good luck. :smile:

**Comment by DanBloomberg** [comment]
Thanks.  Keeping the threshold parameter is fine with me.

**Comment by zdenop** [comment]
Friends, I have 2 requests:
1. Create a new issue / PR for a new topic
2. make diff/patch with `git diff >my_changes.path` (it simplifies integration a lot)

Let's move the discussion about this patch to https://github.com/agl/jbig2enc/pull/107

**Comment by zdenop** [comment]
Regarding the release of the 0.30 version - I am fine with it. I will consider it as a 1.0 prerelease 😁

**Comment by zvezdochiot** [comment]
Resolved: [0.30](https://github.com/agl/jbig2enc/releases/tag/0.30)
